src/cargo/core/compiler/custom_build.rs

```diff
@@ -735,7 +743,8 @@ impl BuildOutput {
                         env.push((key, val));
                     }
                 }
-                "warning" => warnings.push(value.to_string()),
+                // "error" is not parsed here for backwards compatibility, and because this function is for successful outputs
```

Comment:
> I haven't made cargo:error fail the compilation, which remains dependent on the script's exit code. This way this feature is technically backwards-compatible with crates that might have used cargo:error as their custom key-value metadata.

This is referring to

> [cargo:KEY=VALUE](https://doc.rust-lang.org/cargo/reference/build-scripts.html#the-links-manifest-key) — Metadata, used by links scripts.

See https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script

That is a bit of a compatibility trap as theoretically any new directive is a breaking change.

The previous direction from the cargo team was

> Treat the script as having errored whether or not it returns an error code.

I feel like not doing this is a behavior trap for users.  I would recommend re-raising this in the Issue and for us to come up with a compatible solution that allows erroring.

Reply:
I've elaborated on the design here: https://github.com/rust-lang/cargo/issues/10159#issuecomment-1298559078